### PR TITLE
Restore compatibility with GraalVM 25

### LINKED ISSUE #5892: Native build failure since JUnit 6.1.0
I have a simple Java command line application using JUnit 6.0.3 with a classic and native Gradle build.

Since JUnit 6.1.0 './gradlew nativeTest' fails with the following error:

```
Exception in thread "main" com.oracle.svm.core.jdk.UnsupportedFeatureError: SerializationConstructorAccessor class not found for declaringClass: org.junit.platform.engine.UniqueId$SerializedForm (targetConstructorClass: java.lang.Object). Usually adding org.junit.platform.engine.UniqueId$SerializedForm to serialization-config.json fixes the problem.
> Task :nativeTest FAILED
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:121)
gradle/actions: Writing build results to /home/runner/work/_temp/.gradle-actions/build-results/__run-1784609309065.json
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.serialize.SerializationSupport.getSerializationConstructorAccessor(SerializationSupport.java:286)

	at java.base@25.0.3/jdk.internal.reflect.ReflectionFactory.generateConstructor(ReflectionFactory.java:2439)
[Incubating] Problems report is available at: file:///home/runner/work/junit-native-build-failure/junit-native-build-failure/build/reports/problems/problems-report.html
	at java.base@25.0.3/jdk.internal.reflect.ReflectionFactory.newConstructorForSerialization(ReflectionFactory.java:297)
	at java.base@25.0.3/java.io.ObjectStreamClass.getSerializableConstructor(ObjectStreamClass.java:1319)
	at java.base@25.0.3/java.io.ObjectStreamClass.<init>(ObjectStreamClass.java:371)
	at java.base@25.0.3/java.io.ObjectStreamClass$Caches$1.computeValue(ObjectStreamClass.java:93)
	at java.base@25.0.3/java.io.ObjectStreamClass$Caches$1.computeValue(ObjectStreamClass.java:90)
	at java.base@25.0.3/java.io.ClassCache$1.computeValue(ClassCache.java:78)
	at java.base@25.0.3/java.io.ClassCache$1.computeValue(ClassCache.java:75)
	at java.base@25.0.3/java.lang.ClassValue.get(JavaLangSubstitutions.java:542)
	at java.base@25.0.3/java.io.ClassCache.get(ClassCache.java:89)
	at java.base@25.0.3/java.io.ObjectStreamClass.lookup(ObjectStreamClass.java:76)
	at java.base@25.0.3/java.io.ObjectStreamClass.lookup(ObjectStreamClass.java:227)
	at org.junit.platform.engine.UniqueId.<clinit>(UniqueId.java:51)
	at org.junit.platform.engine.discovery.DiscoverySelectors.selectUniqueId(DiscoverySelectors.java:[119](https://github.com/grumpyf0x48/junit-native-build-failure/actions/runs/29802216161/job/88545343306#step:5:120)9)
	at java.base@25.0.3/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
	at java.base@25.0.3/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1716)
	at java.base@25.0.3/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:803)
	at java.base@25.0.3/java.util.stream.ReferencePipeline$7$1FlatMap.accept(ReferencePipeline.java:293)
	at java.base@25.0.3/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
	at java.base@25.0.3/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:197)
	at java.base@25.0.3/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base@25.0.3/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base@25.0.3/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
	at java.base@25.0.3/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
	at java.base@25.0.3/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base@25.0.3/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base@25.0.3/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:723)
	at org.graalvm.junit.platform.NativeImageJUnitLauncher.getSelectors(NativeImageJUnitLauncher.java:169)
	at org.graalvm.junit.platform.NativeImageJUnitLauncher.getTestPlan(NativeImageJUnitLauncher.java:148)
	at org.graalvm.junit.platform.NativeImageJUnitLauncher.main(NativeImageJUnitLauncher.java:[122](https://github.com/grumpyf0x48/junit-native-build-failure/actions/runs/29802216161/job/88545343306#step:5:123))
	at java.base@25.0.3/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)

FAILURE: Build failed with an exception.

```

## Steps to reproduce

[Sample Project](https://github.com/grumpyf0x48/junit-native-build-failure) shows the build error :

- Main branch using JUnit 6.0.3 has GREEN build
- [Pull Request](https://github.com/grumpyf0x48/junit-native-build-failure/pull/1) has RED build

## Context

 - Used versions (Jupiter/Vintage/Platform): 6.1.0
 - Java Version: 25
 - Build Tool/IDE: Gradle

## Workaround

To make the build GREEN, it is possible to add the following content to `src/test/resources/META-INF/native-image/serialization-config.json` :

```
[
  {
    "name": "org.junit.platform.engine.UniqueId$SerializedForm"
  }
]
```

### PR DESCRIPTION
**Pull request by marcphilipp**
- **Add test coverage for GraalVM 25**
- **Log test events and output when using `--info`**
- **Add serialization-config.json for compatibility with GraalVM 25**

Fixes #5892.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://docs.junit.org/snapshot/) and [Release Notes](https://docs.junit.org/snapshot/release-notes.html)

### CODE REVIEW

File: junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[
+  {
+    "name": "org.junit.platform.engine.UniqueId$SerializedForm"
```
Comment (mpkorstanje) [inline junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json]: This looks really fragile. Is there a way to generate it?

File: platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts
Comment (mpkorstanje) [inline platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts:3]: If I understood things right, this plugin also adds reach-ability data. Are we're patching the holes in this plugin?

Comment (mpkorstanje) [reply platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts:3]: And if so, would it make sense for us to generate and distribute that information as a first party?

File: junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json (hunk 1 above)
Comment (marcphilipp) [reply junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json]: I think the [tracing agent](https://www.graalvm.org/latest/reference-manual/native-image/metadata/AutomaticMetadataCollection/#tracing-agent) does that for actual Graal applications. Maybe we can attach that to the test and at least verify it produces the same file?

File: platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts
Comment (marcphilipp) [reply platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts:3]: Doesn't adding it to the jar achieve exactly that?

File: junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json (hunk 1 above)
Comment (marcphilipp) [reply junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json]: @sbrannen @vjovanov Could one of you please chime in here?

File: platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts
Comment (mpkorstanje) [reply platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts:3]: When we do that with this file, are we distributing the complete catalog needed to run JUnit on Graal or just what is needed to patch the information provided by `org.graalvm.buildtools.native`?

Comment (marcphilipp) [reply platform-tooling-support-tests/projects/graalvm-starter/build.gradle.kts:3]: Only what's necessary to run test using this plugin which is the main purpose of running JUnit in native image and I think also where we should draw the line. I'm sure you could implement something using JUnit that wouldn't work but I then you'd have to provide additional metadata on your own.

File: junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json (hunk 1 above)
Comment (marcphilipp) [reply junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json]: I've created an issue asking for feedback:

- https://github.com/graalvm/native-build-tools/issues/1005

Comment (marcphilipp) [reply junit-platform-engine/src/main/resources/META-INF/native-image/org.junit.platform/junit-platform-engine/serialization-config.json]: I've switched to the newer `reachability-metadata.json` now and added a test that uses the tracing agent and verifies that the serializable classes in the shipped file are identical to those generated by the agent for regular (non-native-image) test execution.

**Comment by vjovanov** [comment]
The changes look great!

FYI from the GraalVM side: the 6.1.0 regression this PR fixes is also being caught by the automated compatibility testing in [oracle/graalvm-reachability-metadata](https://github.com/oracle/graalvm-reachability-metadata). The scheduled runs currently fail `native-image` execution on GraalVM 25 for the 6.1.0 artifacts:

- [#9207](https://github.com/oracle/graalvm-reachability-metadata/issues/9207) — `junit-jupiter-engine:6.1.0`
- [#9208](https://github.com/oracle/graalvm-reachability-metadata/issues/9208) — `junit-jupiter-params:6.1.0`
- [#9211](https://github.com/oracle/graalvm-reachability-metadata/issues/9211) — `junit-platform-launcher:6.1.0`
- [#9212](https://github.com/oracle/graalvm-reachability-metadata/issues/9212) — `junit-platform-reporting:6.1.0`

These match the root cause of #5892 (`UniqueId$SerializedForm` requiring serialization metadata), so this PR should resolve them as well.

We're tracking the cleanup on our side in [oracle/graalvm-reachability-metadata#9243](https://github.com/oracle/graalvm-reachability-metadata/issues/9243): once this ships in 6.1.3 with the metadata embedded in `junit-platform-engine`, we'll cap the corresponding entries in the metadata repository so they don't duplicate what the jar provides.

**Comment by mpkorstanje** [comment]
@vjovanov looking at metadata available in `oracle/graalvm-reachability-metadata` I get the impression that in this patch we're adding insufficient meta-data to replace all the metadata provided by `oracle/graalvm-reachability-metadata` for a specific JUnit version. Is that correct? And is that not a problem if you want to cap the version of JUnit for which metadata is supplied?

And since the reach-ability metadata depends on a specific execution of `graalvm/native-build-tools` wouldn't it make more sense for either `graalvm/native-build-tools` or `oracle/graalvm-reachability-metadata` to include a fix?

**Comment by vjovanov** [comment]
Good questions — let me clarify, since "cap" was ambiguous in my earlier comment.

**1. Is the embedded metadata insufficient to replace the repo's metadata?** It isn't a replacement, and it doesn't need to be — the two sources are additive. `native-image` always picks up `META-INF/native-image/**` embedded in jars on the classpath, and native-build-tools supplies the repository metadata *in addition to* that. So with 6.1.3 a user gets both the new serialization registration from the jar and everything the repository already ships.

**2. Doesn't capping strand newer JUnit versions?** No metadata coverage is being removed. The repository's `latest` entries for the JUnit artifacts apply to all newer versions automatically, and they'll keep doing so. The repository never contained the `UniqueId$SerializedForm` registration — that's exactly why our compatibility runs started failing on 6.1.0. The "cleanup" in oracle/graalvm-reachability-metadata#9243 is just: once 6.1.3 is out, our automation records it as tested and we close the failure issues. We'd only ever cap the repository entries if JUnit ends up embedding the complete set (the remaining entries are a handful of conditional JDK-internal registrations and resource globs — we'd be glad to help upstream those too, but it's optional).

**3. Why fix it in JUnit rather than native-build-tools or the metadata repository?** The repository could have carried this fix, but it only reaches builds that use native-build-tools with the metadata repository enabled. Metadata embedded in the jar works for *every* consumer — plain `native-image`, Maven/Gradle native-build-tools, Quarkus, Micronaut, etc. — and is versioned and released in lockstep with the code that needs it (this entry exists because of the 6.1.0 `SerializedForm` change). native-build-tools itself is just the plumbing; it doesn't own per-library metadata. The repository is deliberately a stop-gap for libraries that don't ship their own metadata, so a library adopting the piece that's intrinsic to its implementation — with a test keeping it in sync, as this PR adds — is exactly the direction we encourage.

Also worth noting: we have made JUnit one of our [high-priority projects](https://github.com/oracle/graalvm-reachability-metadata/pull/9245/changes#diff-f26b5ccc312e630a30b707f2b7075e4f11b56ae2272af0e7da38c662d76bdfcaR5) in the metadata repository. Every future JUnit version is checked by our automated compatibility testing as soon as it is released, and any failures are escalated as high-priority tickets and fixed immediately — that is how the 6.1.0 regression was caught in the first place.
